File: Programmers/동굴탐험.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.Queue;

public class 동굴탐험 {
    static int[] check; // 0: start, 1: nothing, 2: end
    static int[] visit; // 0: not visit, 1: visited
    static int[] firstVisit;
    static int[] afterVisit;
    static ArrayList<Integer>[] board;
    static Queue<Integer> q = new LinkedList<>();

    public static void dfs(int num) {
        if(check[num] == 2) {
            afterVisit[num] = 1;
            return;
        }
        
        if(visit[num] == 1) return;
        visit[num] = 1;
        
        for (int i = 0; i < afterVisit.length; i++) {
            System.out.print(visit[i]+" ");
        }
        System.out.println();

        if(check[num] == 0) {
            q.add(firstVisit[num]);
            check[num] = 1;
        }
        for (int i : board[num]) {
            if(visit[i] == 1) continue;
            dfs(i);
        }
    }

    public static boolean solution(int n, int[][] path, int[][] order) {
        //init
        check = new int[n];
        visit = new int[n];
        firstVisit = new int[n];
        afterVisit = new int[n];
        Arrays.fill(check, 1);
        board = new ArrayList[n];
        for (int i = 0; i < n; i++) {
            board[i] = new ArrayList<Integer>();
        }

        //input
        for (int i = 0; i < path.length; i++) {
            board[path[i][0]].add(path[i][1]);
            board[path[i][1]].add(path[i][0]);
        }
        for (int i = 0; i < order.length; i++) {
            firstVisit[order[i][0]] = order[i][1];
            check[order[i][0]] = 0;//from
            check[order[i][1]] = 2;//to
        }
        //search
        dfs(0);
        System.out.print("===\n");

        while(!q.isEmpty()) {
            int x = q.poll();
            check[x] = 1;
            if(afterVisit[x] == 1) dfs(x);
        }
        for (int i = 0; i < n; i++) {
            if(visit[i] != 1) return false;
        }
        return true;
    }
    
    public static void main(String[] args) {
        int n = 9;
        int[][] test1path = {{0,1},{0,3},{0,7},{8,1},{3,6},{1,2},{4,7},{7,5}};
        int[][] test1order = {{8,5},{6,7},{4,1}};
        System.out.println(solution(n, test1path, test1order));
    }
}

File: Programmers/동굴탐험.cpp
#include <string>
#include <vector>
#include <queue>
using namespace std;
//0:먼저 방문할 곳, 2: 0번 방문 후 방문할 곳 1:아무것도아님
int check[200001];
int visit[200001];//방문
int afterVisit[200001];//나중방문
int priorityVisit[200001];//우선방문
vector<int> v[200001];//경로
queue<int> q;
void dfs(int num) {
    if(check[num] == 2) {
        afterVisit[num] = 1;
        return;
    }
    if(visit[num] == 1) return;
    visit[num] = 1;
    if(check[num] == 0) {
        q.push(priorityVisit[num]);
        check[num] = 1;
    }
    for(int next : v[num]) {
        if(visit[next] == 1) continue;
        dfs(next);
    }
}

bool solution(int n, vector<vector<int>> path, vector<vector<int>> order) {
    bool answer = true;
    
    //경로 입력
    for(int i = 0; i < path.size(); i++) {
        v[path[i][0]].push_back(path[i][1]);
        v[path[i][1]].push_back(path[i][0]);
    }
    //방문지 설정
    for(int i = 0; i < n; i++) check[i] = 1;
    for(int i = 0; i < order.size(); i++) {
        check[order[i][0]] = 0;
        check[order[i][1]] = 2;
        priorityVisit[order[i][0]] = order[i][1];
    }
    //사전방문
    dfs(0);
    
    //나중방문
    while(!q.empty()) {
        auto cur = q.front(); q.pop();
        check[cur] = 1;
        if(afterVisit[cur] == 1) dfs(cur);
    }
    
    //방문확인
    for(int i = 0; i < n; i++) {
        if(visit[i] != 1) return false;
    }
    return answer;
}